File: HMMMAssembler/Official/hmmm.go
package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}
func main() {
	//pointer to 1st register
	regs := &node{0, "r1", nil}
	mem := &node{0, "---", nil}
	Stac := &stack{}
	//create registers 1 through 15
	createRegs(regs, "r")
	createRegs(mem, "m")
	if len(os.Args) < 2 {
		reader := bufio.NewReader(os.Stdin)
		fmt.Println("HMMM shell")
		fmt.Println("---------------------")
		var emp []string
		for {
			fmt.Print("ASSEMBLE!> ")
			text, _ := reader.ReadString('\n')
			// convert CRLF to LF
			text = strings.Replace(text, "\n", "", -1)
			text = clean(text)
			if strings.Compare("help", text) == 0 {
				fmt.Println("Any instruction from HMMM is valid except jumps")
				fmt.Println("Type quit or exit to leave the shell")
			}
			if strings.Compare("exit", text) == 0 || strings.Compare("quit", text) == 0 {
				break
			}
			if strings.Compare("avengers", text) == 0 {
				fmt.Printf("Easter egg found. Now go to https://www.youtube.com/watch?v=T5Xeat06R28")
				break
			}
			eval(strings.Split(text, " "), regs, mem, Stac, emp, 0)
		}
	} else {
		//read file to array of strings
		file, err := ioutil.ReadFile(os.Args[1])
		check(err)
		var fileName string = string(file)
		var sA []string = strings.Split(fileName, "\n")
		var strArr []string
		//check for last line being counted this causes serious errors
		//in future make sure that the last line has no whitespace before running the program
		if string(sA[len(sA)-1][0]) == " " || string(sA[len(sA)-1][0]) == "\t" || string(sA[len(sA)-1][0]) == "\n" || string(sA[len(sA)-1][0]) == "\r" {
			strArr = remove(sA, sA[len(sA)-1])
		} else {
			strArr = sA
		}
		var x int = 0
		if string(strArr[0][0]) == "1" {
			x++
		}
		for i := x; i < len(strArr); i++ {
			strArr[i] = clean(strArr[i])
		}
		var s int = 0
		if string(strArr[0][0]) == "0" || string(strArr[0][0]) == "1" {
			s++
		}
		for i := x; i < len(strArr); i++ {
			fmt.Println(strings.Split(strArr[i], " "))
			eval(strings.Split(strArr[i], " "), regs, mem, Stac, strArr, s)
		}
		printList(regs)
		printList(mem)
	}
}
func eval(cmd []string, regs *node, mem *node, s *stack, st []string, shift int) int {
	switch string(cmd[(0 + shift)][0]) {
	case "l":
		if string(cmd[0+shift][4]) == "r" {
			loadr(cmd[1+shift], cmd[2+shift], regs, mem)
		} else {
			loadn(cmd[1+shift], cmd[2+shift], regs, mem)
		}
	case "o":
		or(cmd[1+shift], cmd[2+shift], cmd[3+shift], regs)
	case "x":
		xor(cmd[1+shift], cmd[2+shift], cmd[3+shift], regs)
	case "s":
		if string(cmd[0+shift][1]) == "u" {
			sub(cmd[1+shift], cmd[2+shift], cmd[3+shift], regs)
		} else if string(cmd[0+shift][1]) == "e" {
			setn(cmd[1+shift], cmd[2+shift], regs)
		} else if (string(cmd[0+shift][1]) == "t") && (string(cmd[0+shift][5]) == "r") {
			storer(cmd[1+shift], cmd[2+shift], regs, mem)
		} else {
			storen(cmd[1+shift], cmd[2+shift], regs, mem)
		}
	case "a":
		if strings.Compare("and", cmd[0+shift]) == 0 {
			and(cmd[1+shift], cmd[2+shift], cmd[3+shift], regs)
		} else {
			if strings.Compare("addn", cmd[0+shift]) == 0 {
				addn(cmd[1+shift], cmd[2+shift], regs)
			} else {
				add(cmd[1+shift], cmd[2+shift], cmd[3+shift], regs)
			}
		}
	case "m":
		if strings.Compare("mov", cmd[0+shift]) == 0 {
			copy(cmd[1+shift], cmd[2+shift], regs)
		} else {
			if strings.Compare("mul", cmd[0+shift]) == 0 {
				mul(cmd[1+shift], cmd[2+shift], cmd[3+shift], regs)
			} else {
				mod(cmd[1+shift], cmd[2+shift], cmd[3+shift], regs)
			}
		}
	case "c":
		//calln rX N 	Copy the next address into rX and then jump to mem. addr. N 	call
		if string(cmd[0+shift][1]) == "a" {
			c2, err := strconv.Atoi(cmd[2+shift])
			check(err)
			calln(c2, cmd[1], regs)
			c, err := strconv.Atoi(cmd[2+shift])
			check(err)
			var i int = c
			for i < len(st) {
				eval(strings.Split(st[i], " "), regs, mem, s, st, shift)
				i++
			}
		} else {
			copy(cmd[1+shift], cmd[2+shift], regs)
		}
	case "d":
		divide(cmd[1+shift], cmd[2+shift], cmd[3+shift], regs)
	case "n":
		if strings.Compare("neg", cmd[0+shift]) == 0 {
			neg(cmd[1+shift], cmd[2+shift], regs)
		} //else: nop
	case "e":
		return 2
	case "h":
		return 2
	case "r":
		read(cmd[1+shift], regs)
	case "w":
		write(cmd[1+shift], regs)
	case "p":
		if string(cmd[0+shift][1]) == "u" {
			n := searchList(cmd[1+shift], regs)
			push(s, n.val)
		} else {
			n := searchList(cmd[1+shift], regs)
			n.val = pop(s)
		}
	case "j":
		//jumpn N 	Set program counter to address N 	None
		//jumpr rX 	Set program counter to address in rX 	jump
		if string(cmd[0+shift][1]) == "u" {
			if string(cmd[0+shift][4]) == "n" {
				c, err := strconv.Atoi(cmd[1+shift])
				check(err)
				var i int = c
				for i < len(st) {
					eval(strings.Split(st[i], " "), regs, mem, s, st, shift)
					i++
				}
			} else {
				c := searchList(cmd[1+shift], regs)
				var i int = c.val
				for i < len(st) {
					eval(strings.Split(st[i], " "), regs, mem, s, st, shift)
					i++
				}
			}
		} else {
			/*
				jeqzn rX N 	If rX == 0, then jump to line N 	jeqz
				jnezn rX N 	If rX != 0, then jump to line N 	jnez
				jgtzn rX N 	If rX > 0, then jump to line N 	jgtz
				jltzn rX N 	If rX < 0, then jump to line N 	jltz
			*/
			c := searchList(cmd[1+shift], regs)
			N, err := strconv.Atoi(cmd[2+shift])
			check(err)
			switch string(cmd[0+shift][1]) {
			case "e":
				if c.val == 0 {
					var i int = N
					for i < len(st) {
						eval(strings.Split(st[i], " "), regs, mem, s, st, shift)
						i++
					}
				}
			case "n":
				if c.val != 0 {
					var i int = N
					for i < len(st) {
						eval(strings.Split(st[i], " "), regs, mem, s, st, shift)
						i++
					}
				}
			case "g":
				if c.val > 0 {
					var i int = N
					for i < len(st) {
						eval(strings.Split(st[i], " "), regs, mem, s, st, shift)
						i++
					}
				}
			case "l":
				if c.val < 0 {
					var i int = N
					for i < len(st) {
						eval(strings.Split(st[i], " "), regs, mem, s, st, shift)
						i++
					}
				}
			}

		}
	default:
		fmt.Println("Error: opcode " + cmd[0+shift] + " not found\n")
	}
	return 0
}

File: HMMMAssembler/C/hmmm.c
#include "hmmm.h"
#define LIN 20
int eval(char* s, node* regs, node* mem, stack* a, char* lines[20]);
int main(int argc, char* argv[]){
	stack* st=(stack*)malloc(sizeof(stack));
	stack_init(st);
	node* regs=(node*)malloc(sizeof(node));
	regs=createList(15, 'r');
	node* mem=(node*)malloc(sizeof(node));
	mem=createList(15, 'm');
	char lines[LIN][LIN];
	if(argc<2){
		int i=0;
		while(1){
			printf("ASSEMBLE!>");
			char line[20];
			fgets(line, sizeof(line), stdin);
			strncpy(lines[i++], line, 20);
			eval(line, regs, mem, st, lines);
			displayList(regs);
		}
	}else{
		char const* const fileName=argv[1];
		FILE* file=fopen(fileName, "r");
		char line[20];
		int i=0;
		while(fgets(line, sizeof(line), file)){
			strncpy(lines[i++], line, 20);
		}
		for(int i=0; i<20; i++){
			if(eval(clean(lines[i]), regs, mem, st, lines)==2){
				break;
			}
		}
		//displayList(regs);
		//displayList(mem);
		fclose(file);
	}
	return 0;
}
int eval(char* s, node* regs, node* mem, stack* a, char* lines[20]){
	if(strcmp("N", s)==0){
		return 2;
	}
	char** v=split_line(s);
	switch(v[0][0]){
		case 'l':
			if(v[0][4]=='r'){
				loadr(v[1], v[2], regs, mem);
				break;
			}else{
				loadn(v[1], v[2], regs, mem);
				break;
			}
		case 'o':
			or(v[1], v[2], v[3], regs);
			break;
		case 'x':
			xor(v[1], v[2], v[3], regs);
			break;
		case 's':
			if(v[0][1]=='u'){
				sub(v[1], v[2], v[3], regs);	
			}else if(v[0][1]=='e'){
				setn(v[1], v[2], regs);
			}else if(v[0][1]=='t'&&v[0][5]=='r'){
				storer(v[1], v[2], regs, mem);
			}else{
				storen(v[1], v[2], regs, mem);
			}
			break;
		case 'a':
			if(strcmp("and", v[0])==0){
				and(v[1], v[2], v[3], regs);
			}else{
			if(strcmp("addn", v[0])==0){
				addn(v[1], v[2], regs);
			}else{
				add(v[1], v[2], v[3], regs);
			}
			}
			break;
		case 'm':
			if(strcmp("mov", v[0])==0){
				copy(v[1], v[2], regs);
			}else{
			if(strcmp("mul", v[0])==0){
				mul(v[1], v[2], v[3], regs);
			}else{
				mod(v[1], v[2], v[3], regs);
			}}
			break;
		case 'c':
			copy(v[1], v[2], regs);
			break;
		case 'd':
			divide(v[1], v[2], v[3], regs);
			break;
		case 'n':
			if(strcmp("neg", v[0])==0){
				neg(v[1], v[2], regs);
			} //else: nop
			break;
		case 'e':
			break;
		case 'h':
			return 2;
			break;
		case 'r':
			read(v[1], regs);
			break;
		case 'w':
			write(v[1], regs);
			break;
		case 'p':
			if(v[0][1]=='u'){
				node* n=search(regs, v[1], 'n');
				push(a, n->val);
			}else{
				node *n=search(regs, v[1], 'n');
				n->val=pop(a);
			}
			break;
		case 'j':
			break;
		default:
			printf("Error: opcode ""%s"" not found\n", v[0]);
			break;
	}
	return 0;
}

